src/os/bluestore/BlueFS.cc

```diff
@@ -3744,6 +4113,17 @@ int BlueFS::truncate(FileWriter *h, uint64_t offset)/*_WF_L*/
   vselector->sub_usage(h->file->vselector_hint, h->file->fnode.size - offset);
   h->file->fnode.size = offset;
   h->file->is_dirty = true;
+  // TODO(pere): update wal_size
+  if (h->file->is_new_wal()) {
+    // This assumption comes from reading logs of rocksdb+bluefs where a WAL file follows this pattern:
+    // 1. create wal
+    // 2. open_for_write
+    // 3. close_writer
+    // 4. truncate -> fnode.size
+    // 5. unlink
+    ceph_assert(h->file->fnode.size == offset || offset == 0);
+    h->file->fnode.wal_limit = offset;
```

Comment:
I have a feeling that wal_limit is an "alias" of f->fnode.get_allocated() so it should shrink without pextent list being shrinked. At best it looks like we don't need to persist wal_limit at all - a run-time instance should be sufficient: we recalculate it on space allocation or replay from get_allocated() and that's all.

Reply:
@pereman2 - what do you think? Do we really need to shrink wal_limit here?

Reply:
I think you may be right.

Reply:
well not really. wal_limit signals the range of data that might exist. When we truncate the file we don't remove extents but we do decrease the upper limit represented by wal_limit.

Reply:
OK, so now this looks like an additional(!) mean to limit the scope for the recovery procedure in wal_update_size().
I.e. if no truncate happened - it scans through the space implicitly determined by fnode.allocated() (as wal_limit is equal to the latter after fnode space expansion). And if there was a truncate - it gets a reduced range which is determined by the last truncate offset persisted in wal_limit. Right?
Honestly this looks like an overkill to me - one might just perform scanning through all the space determined by fnode.allocated() and stop it on flush envelope's decode failure.
I can see no real disadvantages due to that but the code would be a bit simpler.

Reply:
I think this is needed because if you truncate and then try to scan `0~allocated` you will find youserlf finding well formed flush extents that seem ok because we never overwrote them.